File: go/pe014.go
package main

import (
	"fmt"
)

func next_term(num uint64) uint64 {
	var retval uint64
	if num%2 == 0 {
		retval = num / 2
	} else {
		retval = num + num + num + 1
	}
	return retval
}

func main() {
	const LAST uint64 = 1000000
	const HLAST uint64 = 500000
	var maxi, max uint64
	for i := HLAST; i < LAST; i++ {
		var counter uint64
		var j uint64
		j = i
		for j != 1 {
			j = next_term(j)
			counter++
		}
		counter++

		if counter > max {
			max = counter
			maxi = i
		}
	}
	fmt.Printf("%d\n", maxi)
}

File: typescript/pe014.ts
function next_term(num:bigint): bigint {
  let retval = 0n
  if (num % 2n) 
    retval = num + num + num + 1n
  else
    retval = num >> 1n
  return retval
}

function pe014(): number {
  const first = 500000n
  const last = 1000000n
  let max = 0n
  let maxi = 0n
  
  for (let i=first; i<last; ++i)
  {
    let counter = 0n
    let j = i
    while (j != 1n)
    {
      j = next_term(j)
      counter++
    }
    counter++
  
    if (counter > max)
    {
      max = counter
      maxi = i
    }
  }
  return Number(maxi)
}

console.log(pe014())
